### PR TITLE
Add Keep in dock option to first run dialog and settings

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/24055

First run dialog:
Screen Shot 2022-10-28 at 8 54 25 AM

Settings: not added to dock
Screen Shot 2022-10-28 at 8 52 18 AM

After added to dock
Screen Shot 2022-10-28 at 8 52 26 AM

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See the linked issue

### CODE REVIEW

File: browser/brave_shell_integration.h [hunk 1]
```diff
@@ -10,6 +10,8 @@
 
 namespace shell_integration {
 
+void PinDefaultShortcutForExistingUsers();
+
```
Comment (sangwoo108) [inline browser/brave_shell_integration.h:14]: How about having `void PinShourcut()` here and unify `win::PinToTaskbar()` and `mac::AddIconToDock()` (leaving definition in _win,mac.cc in compile-time polymorphism manner)? Then we can remove buildflag from many call sites.

```diff
- #if BUILDFLAG(IS_WIN)
-               shell_integration::win::PinToTaskbar();
- #endif
- #if BUILDFLAG(IS_MAC)
-               shell_integration::mac::AddIconToDock();
- #endif
+               shell_integration::PinShortcut();
```

**Review by sangwoo108 — APPROVED** [review]
LGTM with a suggestion

File: browser/brave_shell_integration.h (hunk 1 above)
Comment (simonhong) [reply browser/brave_shell_integration.h:17]: Done - d4f74cd955c95238a8f603235c62a14a5cd7ebbd Seems much better. Thanks for review!

**Comment by simonhong** [comment]
Set as draft as we'll add asking to users to pin(dock).

**Comment by simonhong** [comment]
@sangwoo108 It's ready to review. PTAL again. Issue and descriptions are updated.

File: browser/brave_shell_integration_mac.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace shell_integration::mac {
+
+// No-op if already added.
+void AddIconToDock(
+    base::OnceCallback<void(bool)> result_callback = base::DoNothing());
+void IsIconAddedToDock(base::OnceCallback<void(bool)> result_callback);
+}  // namespace shell_integration::mac
+
+#endif  // BRAVE_BROWSER_BRAVE_SHELL_INTEGRATION_MAC_H_
```
Comment (sangwoo108) [inline browser/brave_shell_integration_mac.h:20]: nit: new line at the EOF

File: browser/brave_shell_integration_mac.mm [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  std::move(result_callback)
+      .Run(dock::ChromeIsInTheDock() == dock::ChromeInDockTrue);
+}
+
+void IsIconAddedToDock(base::OnceCallback<void(bool)> result_callback) {
+  std::move(result_callback)
+      .Run(dock::ChromeIsInTheDock() == dock::ChromeInDockTrue);
+}
+
+}  // namespace shell_integration::mac
```
Comment (sangwoo108) [inline browser/brave_shell_integration_mac.mm:25]: ditto

**Review by sangwoo108 — APPROVED** [review]
Sorry for delayed response. LGTM

File: browser/brave_shell_integration_mac.h (hunk 2 above)
Comment (simonhong) [reply browser/brave_shell_integration_mac.h]: Added.

**Comment by simonhong** [comment]
@rmcfadden3 @rebron Can you check above options' sentence?

File: browser/ui/webui/settings/pin_shortcut_handler.cc [hunk 4]
```diff
@@ -6,15 +6,12 @@
 #include "brave/browser/ui/webui/settings/pin_shortcut_handler.h"
 
 #include "base/bind.h"
+#include "brave/browser/brave_shell_integration.h"
 #include "build/build_config.h"
```
Comment (goodov) [inline browser/ui/webui/settings/pin_shortcut_handler.cc:10]: can be removed?

File: app/brave_generated_resources.grd [hunk 5]
```diff
@@ -932,11 +932,19 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
...
-        </message>
+      <if expr="enable_pin_shortcut">
+        <if expr="is_win">
+          <message name="IDS_FIRSTRUN_DLG_PIN_SHORTCUT_TEXT" desc="Text for pin to taskbar checkbox">
+            Pin to taskbar
+          </message>
+        </if>
+        <if expr="is_macosx">
+          <message name="IDS_FIRSTRUN_DLG_PIN_SHORTCUT_TEXT" desc="Text for pin to taskbar checkbox">
+            Keep in dock
```
Comment (goodov) [inline app/brave_generated_resources.grd:943]: "dock" or "Dock" or "the Dock"? (also applies to other strings)

I believe it should be "the Dock":
https://support.apple.com/guide/mac-help/open-apps-from-the-dock-mh35859/mac
https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/mac/dock.h?q=the%20dock&ss=chromium

File: browser/brave_shell_integration.cc [hunk 6]
```diff
@@ -11,12 +11,39 @@
...
 
+void PinShortcut(Profile* profile,
+                 base::OnceCallback<void(bool)> result_callback) {
+#if BUILDFLAG(IS_WIN)
+  win::PinToTaskbar(profile, std::move(result_callback));
+#elif BUILDFLAG(IS_MAC)
+  // Mac doesn't support profile specific icon in dock.
+  mac::AddIconToDock(std::move(result_callback));
+#elif BUILDFLAG(IS_LINUX)
+  // Not supported on linux yet.
```
Comment (goodov) [inline browser/brave_shell_integration.cc:33]: `NOTIMPLEMENTED()` or `NOTREACHED()`.

File: browser/ui/BUILD.gn [hunk 7]
```diff
@@ -1,3 +1,4 @@
+import("//brave//browser/shell_integrations/buildflags/buildflags.gni")
```
Comment (goodov) [inline browser/ui/BUILD.gn:1]: typo: `//brave//` -> `//brave/`

File: browser/ui/views/brave_first_run_dialog.h [hunk 8]
```diff
@@ -8,6 +8,7 @@
 
 #include "base/callback.h"
 #include "base/memory/raw_ptr.h"
+#include "brave/browser/shell_integrations/buildflags/buildflags.h"
 #include "build/build_config.h"
```
Comment (goodov) [inline browser/ui/views/brave_first_run_dialog.h:12]: can be removed?

File: browser/ui/webui/settings/pin_shortcut_handler.cc (hunk 4 above)
Comment (simonhong) [reply browser/ui/webui/settings/pin_shortcut_handler.cc:10]: Removed.

File: browser/brave_shell_integration.cc (hunk 6 above)
Comment (simonhong) [reply browser/brave_shell_integration.cc:33]: Added `NOTREACHED()`

File: browser/ui/views/brave_first_run_dialog.h (hunk 8 above)
Comment (simonhong) [reply browser/ui/views/brave_first_run_dialog.h:12]: Removed.

File: browser/ui/BUILD.gn (hunk 7 above)
Comment (simonhong) [reply browser/ui/BUILD.gn:1]: Fixed.

File: app/brave_generated_resources.grd (hunk 5 above)
Comment (simonhong) [reply app/brave_generated_resources.grd:943]: Ok, let's wait @rmcfadden3 's review about the labels in settings and dialog.

File: browser/brave_shell_integration.cc (hunk 6 above)
Comment (goodov) [reply browser/brave_shell_integration.cc:34]: can be `NOTREACHED() << "Not supported on Linux yet.";` 😉

**Review by goodov — APPROVED** [review]
lgtm!

File: browser/brave_shell_integration.cc (hunk 6 above)
Comment (simonhong) [reply browser/brave_shell_integration.cc]: Done :) Thanks!

File: app/brave_generated_resources.grd (hunk 5 above)
Comment (rmcfadden3) [reply app/brave_generated_resources.grd:949]: @simonhong — per how Apple does it, it appears it should be "Dock" capitalized, and in the instance of settings copy, the article "the" is not needed (see screenshot below).

So I'd think the copy should be:
- Keep in Dock
- Brave is already in Dock
- 
Screen Shot 2022-10-27 at 2 33 33 PM

Comment (simonhong) [reply app/brave_generated_resources.grd]: @rmcfadden3 Updated. Thanks! 👍🏼

**Comment by stephendonner** [comment]
## Verification `IN-PROGRESS` using 

Brave | 1.47.42 Chromium: 107.0.5304.91 (Official Build) nightly (x86_64)
-- | --
Revision | 3d5948960d62418160796d5831a4d2d7d6c90fa8-refs/branch-heads/5304@{#1097}
OS | macOS Version 11.7.1 (Build 20G918)

#### Prerequisite: make a non-Brave browser the system default via `Apple menu` -> `System Preferences...` -> `General` -> `Default web browser`
<details>

Screen Shot 2022-11-03 at 1 52 02 PM

</details>

#### Case 1: Pin Brave to dock via the `Pin to Dock` checkbox in the first-run dialog - `PASSED`

<details>

1. installed `1.47.42`
2. launched Brave
3. checked the `Keep in Dock` checkbox
4. clicked on the `Set Brave as default` button
5. confirmed I got the `Do you want to change your default web browser to "Brave" or keep using "Firefox"? dialog
6. clicked on `Use "Brave"`
7. confirmed Brave is kept in the dock
8. closed Brave
9. relaunched by clicking on the Brave icon in the dock
10. opened `brave://settings/getStarted`
11. confirmed `Brave is your default browser` text
12. confirmed `Brave is already in Dock` text

#### Confirmed Brave is successfully kept in the dock

step 3 | step 5 | step 7 | steps 11-12
-------|-------|-------|------------
Screen Shot 2022-11-03 at 1 50 26 PM



</details>

#### Case 2: Unpin Brave from the dock and `Pin to Dock` it from `brave://settings/getStarted`

<details>

</details>

#### Case 3: With `Pin to Dock` unchecked (default), click `Set Brave as default` on the first-run dialog

<details>

</details>

#### Case 4: `Pin to Dock` Brave via `brave://settings/getStarted` when a non-Brave browser is set as default via `Apple menu`' -> `System Preferences` -> `General`

<details>

</details>

#### Case 5: Upgrade from `1.45.x (beta)` -> `1.46.x (beta)`
<details>

</details>

**Pull request merged**